WIP: New: capitalized-comments rule (fixes #6055)

platinumazure opened the pull request:
**What issue does this pull request address?**

Implement rule capitalized-comments for JSCS compatibility. (Fixes #6055)

**What changes did you make? (Give an overview)**

Added a new rule "capitalized-comments". (I'm only now realizing that I mistyped the rule name, oops.) If capitalization is required, then rule flags when first non-space non-asterisk character is Unicode lowercase. If capitalization is forbidden, then rule flags when first non-space non-asterisk character is Unicode uppercase. Rule supports overrides for line and block comments and a few other options.

**Is there anything you'd like reviewers to focus on?**

Rule logic should be mostly done. I just need to write tests for some of the other options.

I've borrowed JSCS' letter, uppercase letter, and lowercase letter patterns (these were contributed by [Mathias Bynens](https://github.com/mathiasbynens), who wrote [JavaScript has a Unicode Problem](https://mathiasbynens.be/notes/javascript-unicode) and probably knows what he's doing).

I would appreciate if reviewers could focus on the following:

* Rule schema
* Logic used for cascading and merging options: does it make sense? Is it sufficiently clear?
* Is the "markers" implementation correct? Assumption is that we check the first character of the comment value and compare against the markers.
* JSCS rule specifies the "first word character" as the first non-space, non-asterisk character. I'm wondering if this should be expanded to first non-space, non-punctuation character, but this would cause cases like `// @todo` (currently valid in JSCS' requireCapitalizedComments) to be invalid here.
  * More generally, to what extent are we allowed to deviate from JSCS behavior if JSCS behavior seems arbitrary, questionable, or wrong?

mysticatea commented on conf/eslint.json:
```diff
@@ -140,6 +140,7 @@
         "brace-style": "off",
         "callback-return": "off",
         "camelcase": "off",
+	"capitalized-comments": "off",
```
Seems a tab.

mysticatea commented on lib/rules/capitalized-comments.js:
```diff
@@ -0,0 +1,258 @@
+
+            // 1. Check for markers.
+            if (options.markers.some(m => comment.value[0] === m)) {
+                return;
```
I think it should check this comment with the content without the matched marker.
`spaced-comment` rule is checking spacing preceded by the matched marker.

mysticatea commented on lib/rules/capitalized-comments.js:
```diff
@@ -0,0 +1,258 @@
+                previousToken = sourceCode.getTokenBefore(firstToken);
+
+            return previousToken &&
+                firstToken.loc.line === previousToken.loc.line;
```
Though nit-picking, `previousToken` is not a `boolean` value, but JSDoc says `@returns {boolean}`.

mysticatea commented on lib/rules/capitalized-comments.js:
```diff
@@ -0,0 +1,258 @@
+            // 5. Is the initial word character a letter?
+            const commentWordCharsOnly = comment.value
+                .replace(WHITESPACE, "")
+                .replace(/\*/g, "");        // FIXME: Should this be hardcoded?
```
In `spaced-comment` rule case, the `*` is a marker.

mysticatea commented on lib/rules/capitalized-comments.js:
```diff
@@ -0,0 +1,258 @@
+
+            // 6. Check the case of the initial word character.
+            const isUppercase = UPPERCASE_PATTERN.test(firstWordChar),
+                isLowercase = LOWERCASE_PATTERN.test(firstWordChar);
```
I think we can use standard methods.

```js
const isUppercase = firstWordChar.toLocaleLowerCase() !== firstWordChar,  // if it can be converted to a lowercase character, it's an uppercase character.
    isLowercase = firstWordChar.toLocaleUpperCase() !== firstWordChar;
```

mysticatea commented:
> JSCS rule specifies the "first word character" as the first non-space, non-asterisk character. I'm wondering if this should be expanded to first non-space, non-punctuation character, but this would cause cases like // @todo (currently valid in JSCS' requireCapitalizedComments) to be invalid here.

Sounds good to me.
But we should be careful about the way which distinguishes a character is a punctuator or not. For example, in Japanese, most characters don't have upper/lower case as same as punctuators. And there are wide version in Japanese for most punctuators (e.g. `＠` for `@`).

mysticatea commented:
I'm sorry, I know this option schema is the thing I suggested, but the `overrides` seems too much.

I'd like to rethink the schema, similar to `spaced-comment`:

```
"capitalized-comments": ["error", "always" or "never"]
// or
"capitalized-comments": ["error", {
    "capitalize": "always" or "never",
    "markers": ["*"],
    "ignorePattern": "^\s*(?:eslint|jshint|jslint|istanbul|global|exported|jscs|falls?\s?through)",
    "ignoreInlineComments": false
}]
// or
"capitalized-comments": ["error", {
    "line": {
        "capitalize": "always" or "never",
        "markers": ["*"],
        "ignorePattern": "^\s*(?:eslint|jshint|jslint|istanbul|global|exported|jscs|falls?\s?through)",
        "ignoreInlineComments": false
    },
    "block": {
        "capitalize": "always" or "never",
        "markers": ["*"],
        "ignorePattern": "^\s*(?:eslint|jshint|jslint|istanbul|global|exported|jscs|falls?\s?through)",
        "ignoreInlineComments": false
    }
}]
```

----

```js
const SCHEMA_BODY = {
    type: "object",
    properties: {
        capitalize: {
            enum: ["always", "never"]
        },
        markers: {
            type: "array",
            items: {
                type: "string"
            }
        },
        ignorePattern: {
            type: "string"
        },
        ignoreInlineComments: {
            type: "boolean"
        }
    },
    additionalProperties: false
};
const SCHEMA = [{
    anyOf: [
        {
            enum: ["always", "never"]
        },
        SCHEMA_BODY,
        {
            type: "object",
            properties: {
                line: SCHEMA_BODY,
                block: SCHEMA_BODY
            },
            additionalProperties: false
        }
    ]
}]
```

nzakas commented:
@platinumazure are you planning on finishing this up?

platinumazure commented:
@nzakas Apologies, I've been stuck due to not having a good sense of how much of the original proposal still makes sense (some things such as "markers" were only added because they were in other rules). I need to put together a comment in the issue asking for what I need clarification on.

If I get around to that this week, is that acceptable? Or does this need to be done sooner?

nzakas commented:
@platinumazure that's fine, just didn't want it to languish if you weren't going to finish it up.

platinumazure commented:
@mysticatea I've fixed the minor issues you've highlighted and I've changed the schema per your suggestion. Thanks for your suggestions!

My next step will be to try using `toLocaleLowerCase()` and `toLocaleUpperCase()`, and then add test cases to prove out the schema. After that, I'll return to the issues around markers.

nzakas commented:
@platinumazure this has already been open for two months. What do you think about closing this PR (to get it off the radar) and you can reopen when it's ready?

platinumazure commented:
@nzakas Fine by me. I'm hoping to finish it this week.

On Oct 18, 2016 12:55 PM, "Nicholas C. Zakas" <notifications@github.com>
wrote:

> @platinumazure <https://github.com/platinumazure> this has already been
> open for two months. What do you think about closing this PR (to get it off
> the radar) and you can reopen when it's ready?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eslint/eslint/pull/6946#issuecomment-254587275>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AARWehQ2WjKyroJXwwlGEeDI6MBXD53uks5q1QgAgaJpZM4JpHS5>
> .
>

nzakas commented:
Okay, let's close this for now. When you reopen the PR, it'll appear at the top of the list, so it will get the attention that it deserves.
